packages/cli/src/runners/task-managers/data-request-response-builder.ts

```diff
@@ -0,0 +1,199 @@
+import type {
+	EnvProviderState,
+	IExecuteData,
+	INodeExecutionData,
+	IPinData,
+	IRunData,
+	IRunExecutionData,
+	ITaskDataConnections,
+	IWorkflowExecuteAdditionalData,
+	Workflow,
+	WorkflowParameters,
+} from 'n8n-workflow';
+
+import type { DataRequestResponse, PartialAdditionalData, TaskData } from './task-manager';
+import type { N8nMessage } from '../runner-types';
+
+/**
+ * Builds the response to a data request coming from a Task Runner. Tries to minimize
+ * the amount of data that is sent to the runner by only providing what is requested.
+ */
+export class DataRequestResponseBuilder {
+	private requestedNodeNames = new Set<string>();
+
+	constructor(
+		private readonly taskData: TaskData,
+		private readonly requestParams: N8nMessage.ToRequester.TaskDataRequest['requestParams'],
+	) {
+		this.requestedNodeNames = new Set(requestParams.dataOfNodes);
+
+		if (this.requestParams.prevNode && this.requestParams.dataOfNodes !== 'all') {
+			this.requestedNodeNames.add(this.determinePrevNodeName());
+		}
+	}
+
+	/**
+	 * Builds a response to the data request
+	 */
+	build(): DataRequestResponse {
+		const taskData = this.taskData;
```

Comment:
Nit - We could destructure here for better readability below.

Reply:
Fixed